File: mission/src/main/java/net/tomofiles/skysign/mission/domain/mission/Mission.java
package net.tomofiles.skysign.mission.domain.mission;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@EqualsAndHashCode(of = {"id"})
@ToString
public class Mission {
    @Getter
    private final MissionId id;

    private final Generator generator;
    
    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private String missionName = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Navigation navigation = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version version;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version newVersion;

    Mission(MissionId missionId, Version version, Generator generator) {
        this.id = missionId;
        this.version = version;
        this.newVersion = version;

        this.generator = generator;
    }

    public void nameMission(String name) {
        this.missionName = name;
        this.newVersion = this.generator.newVersion();
    }

    public void replaceNavigationWith(Navigation navigation) {
        this.navigation = navigation;
        this.newVersion = this.generator.newVersion();
    }
}

File: mission/pkg/mission/domain/mission/mission.go
package mission

import (
	"errors"
	"mission/pkg/mission/domain/event"
)

// ID .
type ID string

// Version .
type Version string

const (
	// Original .
	Original = false
	// CarbonCopy .
	CarbonCopy = true
)

var (
	// ErrCannotChange .
	ErrCannotChange = errors.New("cannot change carbon copied mission")
)

// Mission .
type Mission struct {
	id           ID
	name         string
	navigation   *Navigation
	isCarbonCopy bool
	version      Version
	newVersion   Version
	gen          Generator
	pub          event.Publisher
}

// SetPublisher .
func (m *Mission) SetPublisher(pub event.Publisher) {
	m.pub = pub
}

// GetID .
func (m *Mission) GetID() ID {
	return m.id
}

// GetName .
func (m *Mission) GetName() string {
	return m.name
}

// GetNavigation .
func (m *Mission) GetNavigation() *Navigation {
	return m.navigation
}

// GetVersion .
func (m *Mission) GetVersion() Version {
	return m.version
}

// GetNewVersion .
func (m *Mission) GetNewVersion() Version {
	return m.newVersion
}

// NameMission .
func (m *Mission) NameMission(name string) error {
	if m.isCarbonCopy {
		return ErrCannotChange
	}
	m.name = name
	m.newVersion = m.gen.NewVersion()
	return nil
}

// ReplaceNavigationWith .
func (m *Mission) ReplaceNavigationWith(navigation *Navigation) error {
	if m.isCarbonCopy {
		return ErrCannotChange
	}
	m.navigation = navigation
	m.newVersion = m.gen.NewVersion()
	return nil
}

// Generator .
type Generator interface {
	NewID() ID
	NewVersion() Version
}
